rust_src/src/buffers.rs

```diff
@@ -488,6 +486,78 @@ fn assoc_ignore_text_properties(key: LispObject, list: LispObject) -> LispObject
     }
 }
 
+pub enum LispBufferOrName {
+    Buffer(LispBufferRef),
+}
```

Comment:
Shouldn't this enum have a variant for the name?

Reply:
You mean repeat the enum name?

```
pub enum LispBufferOrName {
    LispBufferOrName(LispBufferRef)
}
```

The `Buffer` name is from my initial plan to have `Buffer` and `Name`. But I went with the current code instead.

Reply:
Is this idiomatic rust ?

Reply:
I thought about it some and came up with what I think is a better approach.

If we like this one, I will do something similar with `LispBufferOrCurrent`.

Reply:
I don't think I like this as much as an enum. I had envisoned an enum that held either a string or a buffer:

    enum LispBufferOrName {
        Buffer(LispObject),
        Name(LispObject),
    }

And then two From impls, like this:

    impl From<LispObject> for LispBufferOrName {
        fn from(v: LispObject) -> LispBufferOrName {
            if v.is_string() {
                LispBufferOrName::Name(v)
            } else {
                LispBufferOrName::Buffer(v.as_buffer_or_error())
            }
        }
    }

    impl From<LispBufferOrName> for Option<LispBufferRef> {
        fn from(v: LispBufferOrName) -> Option<LispBufferRef> {
            match v {
                LispBufferOrName::Buffer(b) => Ok(b),
                LispBufferOrName::Name(n) => {
                    cdr(assoc_ignore_text_properties(buffer_or_name, unsafe {
                        Vbuffer_alist
                    })).into()
                }
            }
        }
    }

The first would allow any `lisp_fn` to use `LispBufferOrName` as an argument type, and the second would allow them to trivially convert it into a buffer by calling `.into()`. Neither operation requires any extra allocation, or even calling methods with long names.

Reply:
I tried it. `nsberror` needs the name. So I did this instead. There are also slightly different handlings of nil so I could not easily put the type in LispFn. That was my goal. This was the result.

Reply:
So have an `impl From<LispBufferOrName> for LispBufferRef` which calls `nsberror` instead of returning `None`.

Reply:
yeah, it got messy so I made this simplification. I can recreate the gnarly version so we can talk about it.

Reply:
I also want to work with a 'LispBufferOrCurrent' as well. Using structs this works well. Trying to make From and friends work it was a lot of replicated code and then again, different error handling. Sometimes. Plus, there are currently only a few calls.